core/crypto-api.md

```diff
@@ -0,0 +1,107 @@
+# Crypto API
+
+```typescript
+export declare namespace CryptoTypes {
+  export interface Participant {
+    publicKey: string;
+  }
+
+  export interface KeyPair {
+    privateKey: string;
+    publicKey: string;
+  }
+
+  export interface EncryptParams {
+    message: string;
+    symKey: string;
+    type?: number;
+    iv?: string;
+    senderPublicKey?: string;
+  }
+
+  export interface DecryptParams {
+    symKey: string;
+    encoded: string;
+  }
+
+  export interface EncodingParams {
+    type: Uint8Array;
+    sealed: Uint8Array;
+    iv: Uint8Array;
+    senderPublicKey?: Uint8Array;
+  }
+
+  export interface EncodeOptions {
+    type?: number;
+    senderPublicKey?: string;
+    receiverPublicKey?: string;
+  }
+
+  export interface DecodeOptions {
+    receiverPublicKey?: string;
+  }
+
+  export interface EncodingValidation {
+    type: number;
+    senderPublicKey?: string;
+    receiverPublicKey?: string;
+  }
+
+  export interface TypeOneParams {
+    type: 1;
+    senderPublicKey: string;
+    receiverPublicKey: string;
+  }
+}
+
+export abstract class ICrypto {
+  public abstract name: string;
+
+  public abstract readonly context: string;
+
+  public abstract keychain: IKeyChain;
+
+  constructor(
+    public core: ICore,
+    public logger: Logger,
+    keychain?: IKeyChain
+  ) {}
+
+  public abstract init(): Promise<void>;
+
+  public abstract hasKeys(tag: string): boolean;
+
+  public abstract getClientId(): Promise<string>;
+
+  public abstract generateKeyPair(): Promise<string>;
+
+  public abstract generateSharedKey(
+    selfPublicKey: string,
+    peerPublicKey: string,
+    overrideTopic?: string
+  ): Promise<string>;
+
+  public abstract setSymKey(
+    symKey: string,
+    overrideTopic?: string
+  ): Promise<string>;
+
+  public abstract deleteKeyPair(publicKey: string): Promise<void>;
+
+  public abstract deleteSymKey(topic: string): Promise<void>;
+
+  public abstract encode(
+    topic: string,
+    payload: JsonRpcPayload,
+    opts?: CryptoTypes.EncodeOptions
+  ): Promise<string>;
+
+  public abstract decode(
+    topic: string,
+    encoded: string,
+    opts?: CryptoTypes.DecodeOptions
+  ): Promise<JsonRpcPayload>;
+
+  public abstract signJWT(aud: string): Promise<string>;
+}
```

Comment:
I like the idea of having one place where we agree on the core components but for Android the code snippet above is not very useful. For example we have a separate class for JWT work, and we have more generic methods in regards of setting sym/pub keys. 
Should we just stick only to the general description of what given module is responsible for and drop those api examples?
Whether it makes sense for product apis (Sign, Auth, Chat), in here I'm not quite sure. 
This may also applies for swift team.

Reply:
yeah I agree, on one hand it's good to generalise the concept so everyone knows what we are talking about on the other we already have different implementations so it's difficult to agree on same specs.

Reply:
No worries, I will remove the JWT stuff for now